r/R/dplyr-funcs-datetime.R

```diff
@@ -300,6 +314,34 @@ register_bindings_duration <- function() {
 
     build_expr("cast", x, options = cast_options(to_type = duration(unit = "s")))
   })
+  register_binding("decimal_date", function(date) {
+    y <- build_expr("year", date)
+    start <- call_binding("make_datetime", year = y, tz = "UTC")
+    sofar <- call_binding("difftime", date, start, units = "secs")
+    total <- call_binding(
+      "if_else",
+      build_expr("is_leap_year", date),
+      call_binding("as.integer64", 31622400L),  # number of seconds in a leap year (366 days)
+      call_binding("as.integer64", 31536000L)   # number of seconds in a regular year (365 days)
+      )
+    y + sofar$cast(int64()) / total
+  })
+  register_binding("date_decimal", function(decimal, tz = "UTC") {
+    y <- build_expr("floor", decimal)
+
+    start <- call_binding("make_datetime", year = y, tz = tz)
+    seconds <- call_binding(
+      "if_else",
+      build_expr("is_leap_year", start),
+      call_binding("as.integer64", 31622400L),  # number of seconds in a leap year (366 days)
+      call_binding("as.integer64", 31536000L)   # number of seconds in a regular year (365 days)
+    )
```

Comment:
I think this implementation is fine for now, but we might consider raising a jira for a C++ function that would return the number of seconds in a year. That seems like it might be useful elsewhere, and would let us catch some of the very very edge cases that might happen here (e.g. where we have leap seconds).

Reply:
https://issues.apache.org/jira/browse/ARROW-16066